### ISSUE TITLE
Music crackling/popping

### ISSUE THREAD

**Issue by Novicek** [ihhub/fheroes2#6429]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

Most of the music in this version of the game has a very noticeable crackle, which can be noticed only few seconds into the main menu music for example, but happens during all the music in game I heard so far.

Of course I've tried first if the issue is on my side, but was not able to resolve it: My audio drivers are up to date. I have no audio enhancements on my audio device enabled. My format is set to 48000Hz 24Bits and I've tried almost every possible setting too (reducing channels to 16Bits, changing sample rate to 44100Hz or even 96000Hz).

**However**, switching from my headphones to my speakers, I didn't hear any crackle, so it is most likely on my side. I am still putting this here in case of a suggestion and also because the same music does not have crackle in the other versions of the game (GOG version for DOS/Windows are both fine), so it still possibly has something to do with the way the playback works in this project.

### Save file

Save file not required to reproduce the issue.

### Additional info

Playing the track (homm2_41.ogg) itself via a media player in Windows, I do not hear the crackle either, even with volume set to much higher than in game.

**Comment by Novicek** [comment]
Hey @zenseii, thanks for the reply. I tried what you suggested and the result is still the same, I get crackle when the music plays through fheroes2.

Just to confirm I did it right, I downloaded the zip with flac files, ran the resource_extraction_toolset.bat, dropped the zip in there, which created a music.flac folder. I then backed the original music folder and renamed music.flac to music.

**Comment by Novicek** [comment]
I tested on the snapshot before I posted, but then reverted to 1.0.0. I also experienced with older builds and it's the same everywhere.

I built my PC last January, with a Ryzen9 5900x, RTX 3070, 32GB RAM on the Gigabyte B550 AORUS ELITE motherboard. I don't get performance issues, it's just the music. Most prominent are the town and main menu music, I hear it a lot less on the adventure map and combat screen. Maybe I could record a video to demonstrate.

**Comment by Novicek** [comment]
I just did on my old earbuds. It's still there, though I hear it a bit less.

**Comment by Districh-ru** [comment]
This is also reproducible under Windows. The issue is in SDL2 (or maybe in the way we use it).
When I put the SDL2.dll file from new version (1.0 or latest commit) I get the cracks and also the frequency range of music playback is reduced to 11 kHz (as the game uses 22050 sampling frequency):

![изображение](https://user-images.githubusercontent.com/113276641/210254245-744c71e9-4f33-484f-9368-4b3940108f56.png)

And when I put SDL2.dll file from v.0.9.21 all is OK:
![изображение](https://user-images.githubusercontent.com/113276641/210254396-4ff502bb-ed8b-4637-b77a-e0d2d676f6db.png)

The same problem is for other game sounds. :(

**Comment by Districh-ru** [comment]
In `audio.cpp` we set: `freq = 22050;`.
I've tried to change it to 44100 and this issue has gone...
But checking the original sounds I've found that they have some, maybe mirrored, frequencies. The original sounds sample rate is 22050 Hz, so its frequency range is up to 11025 Hz, and we have frequencies up to 16000 Hz:
![изображение](https://user-images.githubusercontent.com/113276641/210260473-512bb3d9-4dab-49ea-a1a8-c085f4d15cb4.png)
But these 11025-16000 Hz frequencies are about -20 dB, so it may be ignored for now.
Also when I used `freq` 96000 or 192000 Hz I sometimes had cracks in the game sounds, but the music was fine. So the SDL2 re-sampler might have some issues.

@oleg-derevenetz, @ihhub, could you please check this sound playback issue when you have time. Maybe we have to downgrade/upgrade the SDL2 or change the `freq` to 44100. But this may be not good for all platforms.

**Comment by oleg-derevenetz** [comment]
> Maybe we have to downgrade/upgrade the SDL2 or change the freq to 44100. But this may be not good for all platforms.

We can't upgrade it because we already build the latest available release (2.26.1) for Windows. There were [serious issues](https://github.com/libsdl-org/SDL/issues/6391) with their resampler in previous version(s), so I have to downgrade the SDL to 2.0.22 (IIRC), but we can't do this forever. In theory, we can change the frequency to 44100 for Windows only at least temporarily.

**Comment by Districh-ru** [comment]
> In theory, we can change the frequency to 44100 for Windows only - at least temporarily.

This could be a temporary solution. Taking into account that music in OGG files has 48000 Hz sampling rate, IMHO it is better to have this sampling rate in the game. But with current SDL2.dll this sampling rate causes cracks in the game sounds. :(
With 44100 Hz I did not notice any cracks in the sound. This frequency is also good for music files, as the highest frequency in them is about 20.5 kHz.
With SDL 2.0.22 the game always gives me the warning: `Audio::Init:  Audio frequency is initialized as 192000 instead of 22050`. So downgrading the SDL will make the game to use the system sample rate.
Let's hope that in future versions of SDL the resampler will be fixed.

**Comment by oleg-derevenetz** [comment]
Support for the `libsamplerate` was [added recently](https://github.com/microsoft/vcpkg/pull/28554) in the `sdl2` `vcpkg` port, and building `sdl2` with `samplerate` feature may resolve this issue. However `vcpkg` team (Or maybe I will if I succeed) will need to resolve this issue first, because `glib` is mandatory to build the `FluidSynth` plugin that is used by the `SDL_mixer` to play MIDI files.

**Comment by ihhub** [comment]
I have a feeling we are trying to fix issues in SDL library itself :( If this issue is Windows specific should we (possibly not now) use Windows OS libraries to play midi and music files, like DirectMusic?

**Comment by oleg-derevenetz** [comment]
@ihhub

> use Windows OS libraries to play midi and music files, like DirectMusic?

We already tried to use Windows native MIDI (`SDL_mixer` uses it if no FluidSynth plugin or soundfont is present) and we all know that there is a ton of problems with it - it is slow and have issues with volume, that's why we moved to the FluidSynth. Also SDL uses its own resampler implementation (and not the Windows resampler, although it can be used in the SDL code) because it is even worse and brings even more distortion :)

**Comment by oleg-derevenetz** [comment]
Although I was able to build SDL2 with `libsamplerate` support for Windows, I'm not sure how it will work and if `libsamplerate` resampler will be involved at all, because it seems that currently SDL2 uses Windows resampler instead of its own. So let's begin from a simpler approach: just set the audio frequency to `44100` for Windows only.

@Novicek @Districh-ru could you please test the build from #4282 with all supported music types (including OGG, MP3, FLAC and MIDI) as well as in-game sounds for these distortions?

**Comment by Novicek** [comment]
@oleg-derevenetz Tested all sound formats (though MIDI didn't give me crackle even beforehand far as I could tell) and all seems fine.

**Comment by Districh-ru** [comment]
I was busy yesterday, so could not test the build.
I've tested current build and found that the sound clicks are almost inaudible, but they are still present in music (OGG, MP3, FLAC) - in spectrogram they are vertical lines:
![изображение](https://user-images.githubusercontent.com/113276641/210716300-d87e9e91-32ae-4176-bcb1-f3f9bb313099.png)
But this is better, than with 22050 Hz sample rate:
![изображение](https://user-images.githubusercontent.com/113276641/210717248-66739b53-4995-4e53-9c68-f7da88ab1922.png)

And knowing this I'll prefer to use SDL2.dll v. 2.0.22. :(

For long time I've used WinAmp plugin 'YASAPI/NT' (https://sourceforge.net/projects/out-yasapi/), which uses WASAPI. And this plugin with default configuration also had some distortions, including mirrored frequencies. But in plugin configuration there is a setting "SRC Default Quality", enabling which made WinAmp sound perfect - without distortions and mirrored frequencies.
![изображение](https://user-images.githubusercontent.com/113276641/210717728-6eaebc2b-2008-46c1-8ea7-813814512b90.png)

Maybe the authors of SDL should use the same setting to avoid sound distortions of resampler.

**Comment by Districh-ru** [comment]
Hi @oleg-derevenetz, I tried these DLLs and found that there is no difference with the latest master build. Small clicks still present in music.

**Comment by oleg-derevenetz** [comment]
@Districh-ru OK, thanks! I also suspected that there will be no difference, because the use of native WASAPI resampler is currently hardcoded in SDL.

**Comment by oleg-derevenetz** [comment]
@Districh-ru

> Maybe the authors of SDL should use the same setting to avoid sound distortions of resampler.

What this setting does is enable the use of the `AUDCLNT_STREAMFLAGS_SRC_DEFAULT_QUALITY` flag. This flag is already used in SDL:

```c
    /* we've gotten reports that WASAPI's resampler introduces distortions, but in the short term
       it fixes some other WASAPI-specific quirks we haven't quite tracked down.
       Refer to bug #6326 for the immediate concern. */
#if 0
    this->spec.freq = waveformat->nSamplesPerSec;  /* force sampling rate so our resampler kicks in, if necessary. */
#else
    /* favor WASAPI's resampler over our own */
    if (this->spec.freq != waveformat->nSamplesPerSec) {
        streamflags |= (AUDCLNT_STREAMFLAGS_AUTOCONVERTPCM | AUDCLNT_STREAMFLAGS_SRC_DEFAULT_QUALITY);
        waveformat->nSamplesPerSec = this->spec.freq;
        waveformat->nAvgBytesPerSec = waveformat->nSamplesPerSec * waveformat->nChannels * (waveformat->wBitsPerSample / 8);
    }
#endif
```

And still the results are not good.

**Comment by Districh-ru** [comment]
So the issue is in some other place. Let's hope that the developers of SDL could fix it (https://github.com/libsdl-org/SDL/issues/6326, https://github.com/libsdl-org/SDL/issues/6326)

**Comment by oleg-derevenetz** [comment]
> And maybe we should leave this issue opened as the bug still exists.

This bug exists in SDL indeed, but we applied a workaround which works good enough for an ordinary user. However, so be it.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz and @Districh-ru  , as pointed out by a person in Discord it might be that our soundfont is corrupted. This is a reply we got:
```
I redownloaded GeneralUser GS 1.471 from http://www.schristiancollins.com/generaluser.php
and it does not contain the same clicks as fheroes2.sf3
the soundfont itself seem to be corrupted 
```

![image](https://user-images.githubusercontent.com/19829520/226658912-790db31e-84e9-4983-ac73-71575acbaf19.png)

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , the person refers to midi. I will double check with them.

**Comment by oleg-derevenetz** [comment]
@ihhub SF3 format contains samples in some packed format (like MP3) while SF2 format comtains WAV samples. Maybe it is some resampling-related issue. Anyway, I don't hear any artifacts in MIDI mode as well.

**Comment by Districh-ru** [comment]
Hi @ihhub, I tested MIDI play in Windows and Android versions and did not notice any cracks in sound.

**Comment by autoexecbatman** [comment]
But somebody did post an example with cracks present on discord .  https://discord.com/channels/733093692860137523/906187653144268800/1084068036882993212

Also you can just open `fheroes2.sf3` using Polyphone soundfont editor and check for yourself.
Some of the samples compressed in this soundfont contain clicks. 
If it is used to play the midi in game then some notes on some instruments will contain those clicks maybe not audible to every person.

**Comment by oleg-derevenetz** [comment]
Well, I don't hear clicks in SF3 samples. But if someone do, then they are welcome to offer a PR with another soundfont file with appropriate set of instruments (and not very big).

**Comment by oleg-derevenetz** [comment]
@autoexecbatman I don't have a Discord account and do not participate in fheroes2 Discord channel. I just don't like all these messengers because they mostly distract me from thinking and work :)

**Comment by autoexecbatman** [comment]
@oleg-derevenetz Oh ok . 
Anyway long story short. 
http://www.schristiancollins.com/generaluser.php
[GeneralUser GS 1.471] - soundfont does not contain these clicks . 
I think it is the same soundfont that `fheroes.sf3` was made of .

**Comment by oleg-derevenetz** [comment]
@autoexecbatman Earlier in this issue I mentioned that SF3 format is different from SF2, because samples are stored in compressed form in SF3. Original SF2 file is too big even if all unused instruments are removed from it.

**Comment by Districh-ru** [comment]
@ihhub, @autoexecbatman, @oleg-derevenetz,
After some tests in the headphones I noticed the cracks and clicks when the music volume is set to 10. Previously I checked it at 7.
IMHO - it is a bug of simple resampler, which tries to interpolate the sound wave and sometimes the interpolation gives the value more than the maximum sample amplitude and if the Volume is already set to maximum then such value is clipped.
And some "super smart" sound cards do not only clip the sound, but also "to prevent speakers from damage" sets the value of the wave to 0 and we get the pulse from the maximum amplitude to zero. I've encountered such "expensive" ADC at my work.
So to prevent possible sound clipping  and to reduce the THD I never set the volume more then 90-95% of maximum.

The Polyphone soundfont editor also gives me clicks at the maximum amplitude of samples. But OpenMPT, which can also play the samples from soundfonts, does not produce clicks...

**Comment by autoexecbatman** [comment]
@Districh-ru 
![image](https://user-images.githubusercontent.com/91700559/226727969-9f67c429-d36a-4ef7-93a6-e132b4d84047.png)
The clicks are burned inside the samples .

**Comment by autoexecbatman** [comment]
I converted again like I said .
from `*.sf2` to `*.sf3` removing all the instruments to match `fheroes2.sf2`
And now it produces clicks !

**Comment by autoexecbatman** [comment]
The problem is converting from `*.sf2` to `*.sf3` without introducing clicks.

**Comment by Districh-ru** [comment]
> @Districh-ru 
The clicks are burned inside the samples .

@autoexecbatman I don't believe that Polyphone soundfont editor is a very perfect software and it may have bugs in its decoder (It gave me a lot of clicks). So I also used OpenMPT to check soundfont and it gives this:
![изображение](https://user-images.githubusercontent.com/113276641/226733395-c45e4324-93f9-4b4b-a7e1-6e1fc341fba4.png)

The soundfont is OK, but it contains some loud instruments and MIDI mixer does not correctly fix the overload and simply clips the sound.
For comparison (the same game, the same map, all settings the same except music volume):
 Volume8: 

https://user-images.githubusercontent.com/113276641/226736957-ac061bf3-b091-42a9-a529-4c5fc593cde7.mp4

Volume 10:

https://user-images.githubusercontent.com/113276641/226737188-03269fe4-e334-417b-a1d2-c4a706b3fcec.mp4

Maybe a possible solution for this issue could be to reduce the MIDI volume.

The other solution could be - reduce the volume of samples in SF3 file. :)

**Comment by Districh-ru** [comment]
After some tests I can consider that the clicks in MIDI playback are produced by SDL midi synthesizer and/or SDL sound mixer. (It makes midi music too loud).
I put **GeneralUser GS v1.471.sf2** instead of **fheroes2.sf3** in the 'files\soundfonts\' folder and got the same clicks at 10 music volume.
Then I deleted all soundfonts from fheroes2 to make it use the system midi playback and configured *CoolSoft VirtualMIDISynth* (which I always use to play midi files) to play sounds from **GeneralUser GS v1.471.sf2**, started fheroes2 and did not get any clicks in midi music with volume 10.

https://user-images.githubusercontent.com/113276641/226968208-2931eccc-1264-4cd2-b271-692510f08aa8.mp4

@autoexecbatman, could you please test, do you experience clicks in midi music when music volume is set to 8 with original **fheroes2.sf3**?

**Comment by autoexecbatman** [comment]
Using: `fheroes.sf3`(i think it should use this soundfont unless it somehow uses something else). 
Build: `Debug-SDL1`. OS: `win10 x64`
In-game I get no clicks and I hear no clipping at max volume in-game.
The music sounds different from yours and my signal is much colder. (got some clicks due screen recorder)

https://user-images.githubusercontent.com/91700559/227075398-94c61293-8cf7-4954-ae6e-8c85c8f8127e.mp4

...
I am curious how another user got clicks in-game ?

**Comment by autoexecbatman** [comment]
https://user-images.githubusercontent.com/91700559/227081127-791b38ca-7b92-4443-8720-9b6543a7f1bb.mp4

`*.sf2` files are used in this example here.
The same happens actually if I open it with Polyphone using `*.sf3` or exporting the samples to `*.wav` using Polyphone.
So the clicks are resolved ? 
If something clips just turn down the overall gain.

**Comment by autoexecbatman** [comment]
https://user-images.githubusercontent.com/91700559/227085771-26254755-d381-4360-b32e-d32e0b869623.mp4

**Comment by Districh-ru** [comment]
Hi @autoexecbatman, many thanks tor testing!

The decoding of .sf3 have some bugs, as we can hear.But in fheroes2 it decodes correctly as we have no clicks with low music volume.

I think that the cause of midi music clipping is the hardcoded 1.2 gain in SDL_mixer:
```cpp
static void FLUIDSYNTH_SetVolume(void *context, int volume)
{
    FLUIDSYNTH_Music *music = (FLUIDSYNTH_Music *)context;
    /* FluidSynth's default is 0.2. Make 1.2 the maximum. */
    music->volume = volume;
    fluidsynth.fluid_synth_set_gain(music->synth, (float) (volume * 1.2 / MIX_MAX_VOLUME));
}
```
https://github.com/libsdl-org/SDL_mixer/blob/48608bf6507234c3f434ddbb2931f2ead0df9668/src/codecs/music_fluidsynth.c#L259-L265

As we cannot change this hardcoded value we can reduce the volume for midi playback by 1.2. :)

I'll try to fix this possible midi clipping issue in the next few days.

**Comment by oleg-derevenetz** [comment]
> As we cannot change this hardcoded value we can reduce the volume for midi playback by 1.2. :)

Please note that fheroes2 should (ideally) be backend-agnostic. SDL_Mixer has several MIDI backends, and FluidSynth is just one of them. Even on Windows, if we remove the soundfont file, the native Windows MIDI (with its own issues) will be used. What can we say about builds for other systems, where we have no control at all and do not know which backend is used by SDL_Mixer? We should not change our code to please a particular MIDI backend of several available ones.

**Comment by Districh-ru** [comment]
Hi @oleg-derevenetz, thanks for warning.
Before starting to change out code I wanted to figure out can we determine what backend is used by SDL_mixer.
SDL_mixer already has issue https://github.com/libsdl-org/SDL_mixer/issues/284 in which it is mentioned that FLuidSynth midi music is too loud. But for 3 years this issue has not been addressed. :( Maybe because it focuses on changing default playback settings and the loudness of playback is mentioned only in addition.
I'll try to open a new issue in SDL_Mixer repo.

Yes, it is not good to make our engine adopt to every backend (especially if we don't know what backend it is). But currently we can wait until SDL_mixer will set FluidSynth gain to 1 and tell all fheroes2 users not to use music volume more than 8 for midi playback. Or try to make a temporary patch if it is possible.

I will not make changes to our engine if there is no correct way to check if FluidSynth is in use.

**Comment by Districh-ru** [comment]
Also I think we have limit the maximum sound volume to make a room for resampler, as after the resamle some new sample bins can have values more than maximum (or minimum) allowed.
In the next example if we increase sample rate by 2 and the resampler dos not have a magnitude limiter than we will have a reading that will be clipped:
![изображение](https://user-images.githubusercontent.com/113276641/227559023-9d44427e-e365-4cc8-b0f3-42e4a5ba81ff.png)

This is just an assumption that I want to test soon

**Comment by oleg-derevenetz** [comment]
> But for 3 years this issue has not been addressed. :( Maybe because it focuses on changing default playback settings and the loudness of playback is mentioned only in addition.
I'll try to open a new issue in SDL_Mixer repo.

Usually pull requests gain more attention than just issues. If you offer a pull request with the fix, it might be addressed faster. Also in theory we can apply our own patch when building Windows packages in the [fheroes2-prebuilt-deps](https://github.com/fheroes2/fheroes2-prebuilt-deps) repo, but again - it will work only for our own Windows distribution, but not for 3rd-party builds/packages.

**Comment by Districh-ru** [comment]
> For Android builds there is different backend used by SDL_mixer (Timidity) and on Linux this depends on build settings of SDL_mixer package for the corresponding Linux flavor/distribution. Some may use Timidity and some may use FluidSynth. We have no way to control this.

This is bad news that we can not get what backend is used for midi playback. And cannot adjust its settings.
I jus tested Android version (Timidity) and foud that it has clicks in samples, decoded from `.sf3`, when playing MIDI like in videos by @autoexecbatman  :(

**Comment by Districh-ru** [comment]
Ubuntu version experience the same clicks in MIDI music as Widows version if music volume is more than 8.

**Comment by oleg-derevenetz** [comment]
> I just tested Android version (Timidity) and foud that it has clicks in samples, decoded from `.sf3`

Timidity does not support SF3, it uses its own samples (GUS patches) taken from here:

https://github.com/libsdl-org/SDL_mixer

(see the link in the README file).

**Comment by Districh-ru** [comment]
Android version sounds a little different so I was wrong saying about `.sf3`. But the clicks are there for all music volumes.
The Ubuntu version sounds the same as Windows version so I think that FluidSynth is used there.
This issue becomes more complex...

**Comment by oleg-derevenetz** [comment]
> And cannot adjust its settings.

I still believe that per-backend adjustments should be performed on SDL_mixer side. This should be transparent for client apps (like fheroes2) because that's the whole point of SDL_mixer-like libs :)

**Comment by Districh-ru** [comment]
I agree that, but then it is needed that all backends to act the same way (by default): don't have extra gain or extra reverberation or something else.
And I think that there always should be an ability to manually set some parameters. This automated backend switch is a black box: we tell it to play midi and don't know how the real result: use system midi or FluidSynth/Timidity. Or this black box has to just simply say (by demand) what it is doing (what backend it is using and with what parameters).

**Comment by oleg-derevenetz** [comment]
> Why we can not get the type of backend and its real volume gain the same way. :)

Because this type of backend may be simply unsupported or buggy on the target platform (like native MIDI on Windows), and because of this the package maintainer decided to build SDL_mixer with a different backend. You will ask for non-existent backend, and what's then? :) Then the app initialization will just fail.

**Comment by Districh-ru** [comment]
> IMHO the less the client code has to think about choosing and configuring backends, the better :)

Absolutely agree!

> You will ask for non-existent backend, and what's then? :)

I just want to ask what backend is set and don't want to force set it :)
Without knowing what backend is used currently I cannot make a temporary patch to reduce MIDI volume when FluidSynth is used.

It turns out that I hastened with the conclusion that I could quickly solve MIDI clicks issue... :(

**Comment by oleg-derevenetz** [comment]
> Without knowing what backend is used currently I cannot make a temporary patch to reduce MIDI volume when FluidSynth is used.

Since, as you say, these clicks are also observed with other brands (at least with Timidity on Android), they are probably caused by something more general than just FluidSynth backend volume settings.

**Comment by Districh-ru** [comment]
> Since, as you say, these clicks are also observed with other backends (at least with Timidity on Android), they are probably caused by something more general than just FluidSynth backend volume settings.

With FluidSynth the clicks only take place when music volume is set to 9 or 10, with Timidity (Android) the clicks present at every music volume. Tomorrow I'll do some more tests on other Android device.

**Comment by autoexecbatman** [comment]
@oleg-derevenetz I wonder why there is a size constraint that enforces the usage of `*.sf3` format ?

**Comment by oleg-derevenetz** [comment]
> I wonder why there is a size constraint that enforces the usage of *.sf3 format ?

Why do you think that SF3 is problem in the first place? According to @Districh-ru it is not, and the same issue is observed with SF2 on high audio volumes. If the result is the same, why we need to ship ~30Mb instead of ~2.5Mb?

**Comment by autoexecbatman** [comment]
Using `GeneralUser GS v1.471.sf2` I did not get the same clicks. 
in short it could be a compression issue. So if there is a way to check if using `GeneralUser GS v1.471.sf2` on android solves something.

**Comment by oleg-derevenetz** [comment]
> So if there is a way to check if using GeneralUser GS v1.471.sf2 on android solves something.

OMG guys. Android does not use SF2/SF3 soundfonts at all. It uses Timidity backend which uses musical instruments samples from [GUS patches](https://github.com/ihhub/fheroes2/tree/master/android/app/src/main/assets/instruments). And according to @Districh-ru there are also "clicks".

**Comment by autoexecbatman** [comment]
In that case I am out of ideas. Could just set a warning that it is ancient tech and could produce unwanted artifacts when used.
(if not done already)
I got a bug also in GOG version when the music goes out of tune and plays in a different pitch.

**Comment by Districh-ru** [comment]
Hello @autoexecbatman, I unpacked the samples from **fheroes2.sf3** (using RIFF and SF format descriptions) and checked them using Audacity and found no clicks in the samples, except the *Applause.ogg* which contains claps - so I can conclude that **fheroes2.sf3** it is OK.

Here are the unpacked samples if you want to check them: [unpacked.zip](https://github.com/ihhub/fheroes2/files/11129992/unpacked.zip)

**Comment by lothar-cell** [comment]
So, after reading through the thread, I still don't know how to fix the cracking issue (MacOS/Wine). Tried placing the FLAC files from GOG in the MUSIC folder, and the latest .ogg in unpacked.zip.

Can someone explain to a user how to fix the cracking audio?

**Comment by oleg-derevenetz** [comment]
Hi @lothar-cell

> So, after reading through the thread, I still don't know how to fix the cracking issue (MacOS/Wine).

Is there any specific reason why you use such an exotic configuration instead of running fheroes2 directly on macOS?

**Comment by lothar-cell** [comment]
I couldn't make the MacOS version work with the .sh script and I don't know how to/don't want to build from source, and Wine with the .exe works out of the box.

**Comment by oleg-derevenetz** [comment]
> I couldn't make the MacOS version work with the .sh script

What exactly do you mean? Where did you install fheroes2 from (MacPorts, Homebrew?) and what exactly is not working?

**Comment by lothar-cell** [comment]
1. downloaded the MacOS version from there: https://github.com/ihhub/fheroes2/releases/download/1.0.12/fheroes2_macos_x86-64_SDL2.zip
2. double click on install_sdl2.sh - Terminal launches in the background, nothing happens
3. dragged the script onto the Terminal window:
<img width="682" alt="Screenshot 2024-02-10 at 10 15 59" src="https://github.com/ihhub/fheroes2/assets/2356354/eaac37fa-6e2a-4c60-b1db-5f8dafe36891">
4. gave up and installed the win64 version with Wine straight away without problems

**Comment by lothar-cell** [comment]
However, does this have anything to do with the sound issue? As I read it's a problem on all platforms.

**Comment by oleg-derevenetz** [comment]
> 1. downloaded the MacOS version from [here](https://github.com/ihhub/fheroes2/releases/download/1.0.12/fheroes2_macos_x86-64_SDL2.zip)

Have you tried installing fheroes2 using Homebrew or MacPorts? They will install all the necessary dependencies.

> However, does this have anything to do with the sound issue? As I read it's a problem on all platforms.

I do not know where you got this from. There are several _different_ issues observed by _different_ people in _different_ conditions - some issues are caused by the operation of the resampler on a certain platform (sampling rate was tuned at the time because of this), some occur only at maximum volume and only with MIDI music. Nobody knows what's going on with the audio "drivers" and the resampler in Wine.

**Comment by lothar-cell** [comment]
> Have you tried installing fheroes2 using Homebrew or MacPorts? They will install all the necessary dependencies.
I installed it with Crossover, not using MacPorts, where can I find instructions for Homebrew?

**Comment by oleg-derevenetz** [comment]
> where can I find instructions for Homebrew?

In our installation guide:

https://ihhub.github.io/fheroes2/INSTALL.html#homebrew-mac

**Comment by lothar-cell** [comment]
Thank you, I managed to install it and the audio issue is gone:)
But man, this was complicated, installs in a hidden folder to /usr/local/Cellar/fheroes2/1.0.12/

Can I move this whole Cellar folder from there, as it is 3.43GB?

**Comment by oleg-derevenetz** [comment]
> But man, this was complicated, installs in a hidden folder to /usr/local/Cellar/fheroes2/1.0.12/
> Can I move this whole Cellar folder from there, as it is 3.43GB?

The path to the installed apps is determined by package manager (Homebrew). I personally don't use Homebrew, I use MacPorts, but I suppose that you can try to move it to other place and create a symlink to the new location (sort of `sudo ln -s /some/new/location/Cellar /usr/local/Cellar`). Try to ask this question in the Homebrew [support forum](https://github.com/orgs/Homebrew/discussions).

**Comment by lothar-cell** [comment]
Thanks @oleg-derevenetz !

**Comment by oleg-derevenetz** [comment]
I am closing this issue because it has, in fact, turned into a set of unrelated issues and only confuses people.

Original issue with SDL resampler on Windows should have been mitigated by #6455, and then there were additional fixes in the SDL itself related to resampling.

SDL_mixer issue with sound artifacts when playing MIDI music with max volume should have been fixed by https://github.com/libsdl-org/SDL_mixer/pull/508 (credits to @Districh-ru) and this fix should be included into SDL_mixer 2.8.0, which is now used by our Windows and Android dependencies starting from #8381.

For new issues of a similar nature (if they appear), I recommend opening separate issues so as not to mix everything into one pile.
